core/src/saros/git/JGitFacade.java

```diff
@@ -0,0 +1,190 @@
+    Ref actualRef;
+    try {
+      actualRef = repo.exactRef(actual);
+    } catch (IOException e) {
+      throw new IOException("failed while resolved actual", e);
+    }
+    if (actualRef == null)
+      throw new IllegalArgumentException("failed while resolved actual: not existing ref");
+
+    bundlewriter.include("refs/heads/bundle", actualRef.getObjectId());
+
+    // The second information is the commit to start with
+    if (basis != null && !basis.trim().isEmpty()) {
+      try {
+        RevWalk walk = new RevWalk(repo);
+
+        RevCommit basisCommit = walk.parseCommit(repo.resolve(basis));
+
+        walk.close();
+
+        bundlewriter.assume(basisCommit);
+      } catch (NullPointerException | IOException e) {
+        throw new IOException("failed while resolved basis", e);
+      }
+    }
+
+    // After finding the range of commits in the previous steps the bundle object will be written.
+    NullProgressMonitor monitor = NullProgressMonitor.INSTANCE;
+    ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
+
+    bundlewriter.writeBundle(monitor, byteArrayOutputStream);
+
+    return byteArrayOutputStream.toByteArray();
+  }
+
+  /**
+   * Fetching from a bundle to a Git repo
+   *
+   * @param bundle See <a href="https://git-scm.com/docs/git-bundle">git bundle</a>
+   * @throws IllegalArgumentException {@code workDirTree} or {@code bundle} is null and can't be
+   *     resolved
+   * @throws IOException failed while read from {@code workDirTree}, {@code bundle} handling or
+   *     while fetch
+   * @throws URISyntaxException
+   */
+  public void fetchFromBundle(byte[] bundle)
+      throws IllegalArgumentException, IOException, URISyntaxException {
+
+    Git git = getGit();
+
+    if (bundle == null) throw new IllegalArgumentException("bundle is null and can't be resolved");
+
+    try {
+      RemoteRemoveCommand remoteRemoveCommand = git.remoteRemove();
+      remoteRemoveCommand.setName("bundle");
+      remoteRemoveCommand.call();
+    } catch (GitAPIException e) {
+      throw new IOException("failed to remove old bundle path", e);
+    }
+
+    final URIish uri = new URIish("in-memory://");
+    final ByteArrayInputStream in = new ByteArrayInputStream(bundle);
+    final RefSpec rs = new RefSpec("refs/heads/*:refs/heads/*");
+    final Set<RefSpec> refs = Collections.singleton(rs);
+
+    try (TransportBundleStream transport =
+        new TransportBundleStream(git.getRepository(), uri, in)) {
+      transport.fetch(NullProgressMonitor.INSTANCE, refs);
+    } catch (org.eclipse.jgit.errors.TransportException e) {
+      throw new IOException("fetch failed", e);
+    }
+  }
+
+  /**
+   * @return The URL to access/address directly the .git directory
+   * @throws IllegalArgumentException {@code workDirTree} is null can't be resolved
+   * @throws IOException failed while read
+   */
+  String getUrl() throws IllegalArgumentException, IOException {
```

Comment:
could be private?

Reply:
I just need this method in test code. I placed it there. done